planner/core/prepare_test.go

```diff
@@ -748,3 +748,36 @@ func (s *testPrepareSuite) TestPrepareForGroupByMultiItems(c *C) {
 	tk.MustExec(`prepare stmt2 from "select sum(b) from t group by ?, ?"`)
 	tk.MustQuery(`execute stmt2 using @v1, @v2`).Check(testkit.Rows("10"))
 }
+
+func (s *testPrepareSuite) TestInvisibleIndex(c *C) {
+	defer testleak.AfterTest(c)()
+	store, dom, err := newStoreWithBootstrap()
+	c.Assert(err, IsNil)
+	tk := testkit.NewTestKit(c, store)
+	defer func() {
+		dom.Close()
+		store.Close()
+	}()
+
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a int, unique idx_a(a))")
+	tk.MustExec("insert into t values(1)")
+	tk.MustExec(`prepare stmt1 from "select a from t order by a"`)
+
+	tk.MustQuery("execute stmt1").Check(testkit.Rows("1"))
+	tk.MustQuery("execute stmt1").Check(testkit.Rows("1"))
+	c.Assert(len(tk.Se.GetSessionVars().StmtCtx.IndexNames), Equals, 1)
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.IndexNames[0], Equals, "t:idx_a")
+
+	tk.MustExec("alter table t alter index idx_a invisible")
+	tk.MustQuery("execute stmt1").Check(testkit.Rows("1"))
+	tk.MustQuery("execute stmt1").Check(testkit.Rows("1"))
```

Comment:
Why this statement needs to be executed twice?

Reply:
https://github.com/pingcap/tidb/blob/014d101a5d7fbd78f4bc306251181c01c0f01c7c/planner/core/prepare_test.go#L74-L76
we need to execute sql at least twice to test the prepare statement cache.

Reply:
Got it!

Reply:
Got it!